bin/DataDownloadAndMerge.pm

```diff
@@ -0,0 +1,195 @@
+"======================================================================\n";
+    print STDERR
+"    Step4+5: Merge/Downloading from HDFS file system to local directory \n";
+    print STDERR
+"======================================================================\n";
+
+    ################### Time start #########################
+    my $start_time       = time();
+    my $time_stamp_start = Utility::get_time_stamp();
+    print STDERR "\nStarted at:  $time_stamp_start\n";
+
+    ### Paired-end data download
+    $self->paired_end_data($args_dict);
+
+    ### Single-end data download
+    $self->single_end_data($args_dict);
+
+    ################### Time end #########################
+    my $end_time       = time();
+    my $execution_time = Utility::get_execution_time( $start_time, $end_time );
+    my $time_stamp_end = Utility::get_time_stamp();
+    print STDERR "Finished at: $time_stamp_end\n";
+    print STDERR "Duration: $execution_time\n\n";
+}
+
+sub paired_end_data {
+    my ( $self, $args_dict ) = @_;
+
+    $args_dict->{"read_folder"}       = "fastq_paired_end";
+    $args_dict->{"read_type"}         = "pe";
+    $args_dict->{"final_output_file"} = "DistMap_output_Paired_end_reads";
+
+    my $hadoop_exe    = $args_dict->{'hadoop_exe'};
+    my $hdfs_exe      = $args_dict->{"hdfs_exe"};
+    my $readtools     = $args_dict->{"readtools"};
+    my $output        = "/$args_dict->{'random_id'}" . "_output";
+    my $output_folder = "$output/$args_dict->{'read_folder'}" . "_mapping";
+
+    my $file_count = $self->get_file_list(
+"$args_dict->{'output_directory'}/$args_dict->{'random_id'}/$args_dict->{'read_folder'}"
+    );
+
+    my $hdfs_input_folder =
+      "/$args_dict->{'random_id'}" . "_input/fastq_paired_end";
+
+    my @cmd = ();
+    if ( ( system("$hdfs_exe dfs -test -d $hdfs_input_folder") == 0 ) ) {
+        @cmd = `$args_dict->{'hdfs_exe'} dfs -ls $hdfs_input_folder`;
+    }
+
+    my $hdfs_file_count = 0;
+
+    if ( scalar(@cmd) > 1 ) {
+        $hdfs_file_count = $cmd[0];
+        $hdfs_file_count =~ s/\D//g;
+    }
+
+    if ( $file_count > 0 or $hdfs_file_count > 0 ) {
+
+        my $i = 0;
+        foreach my $mapper_name ( @{ $args_dict->{"mapper"} } ) {
+            my $mapper_path = $args_dict->{"mapper_path"}->[$i];
+            my $mapper_args = $args_dict->{"mapper_args"}->[$i];
+
+            $args_dict->{"final_output_file"} =
+              lc($mapper_name) . "_DistMap_output_Paired_end_reads";
+
+            my $output_folder =
+                "$output/$args_dict->{'read_folder'}"
+              . "_mapping_"
+              . lc($mapper_name);
+            my $local_output_dir = $args_dict->{"output_directory"};
+            my $output_file =
+                $args_dict->{"final_output_file"} . "."
+              . $args_dict->{"output_format"};
+            my $download_command =
+"JAVA_OPTS=-Xmx8g -Dsnappy.disable=true $readtools --input $output_folder --output $local_output_dir/$output_file";
```

Comment:
ad 1+2, see below

ad --TMP_DIR, how much space does that need? Normally, tmp_dirs should be created in /var/tmp/

Reply:
Some people already told me that they have problems using a download tool, because the temp directory is too small (I do not know why that's the case).

The temp directory in the download task is used for sorting on the fly -  it is similar to the Picard's SortSam, which is generating small pseudo-bam files that are pre-sorted, and once the work is finished it merges them in the sorted way. Thus, I believe that it requires a similar amount of space as the final BAM (a bit less, because it is compressed in a different way).

The problem is that the snappy library is disabled, but that was an old bug in the library that I was using that was kind of solve (now, almost all architectures should work without that param). We should evaluate if snappy can be enabled in our computers to change that, but in a different PR.

Reply:
It seems to me that it would be difficult to predict where that tmp stuff should go, because space availability may change from vetgrid to vetgrid. The better solution may be to allow an optional --merge-tmp-dir argument that is just passed through to readtools. Should be easy, I'll send a commit presently.

Reply:
I agree that a param is the best way to go, but I believe that the default should be in the output directory (the capacity should be enough anyway for running the download, if not it will fail anyway) and with a hidden folder (e.g., `.distmap_download_merge_tmp`).

In addition, I checked my code and apart of the sorting tmp files, the batches are stored in the tmp directory. So it is a lot of data for a normal tmp directory created by java, I believe.

Reply:
Alternatively, the command line argument could simply set/reset TMPDIR, so that the specified folder would be used whenever temporary files required. Would be simpler, but I did not do it because I am not sure which parts of the code would be affected.